Feat: Allow macros in python model properties

themisvaltinos opened the pull request:
This update enables macros to be passed in the properties for Python models, aside from `audits`, `signals`, `merge filter` and `physical_properties` which are rendered at evaluation time (the latter will be addressed on a subsequent pr). This allows to dynamically control for example the `enabled` flag based on the gateway:

```python
@model(
  ...,
  enabled="@IF(@gateway = 'local', True, False)",
)
```

izeigerman commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
+                    )
+                )
+
+        renderer_kwargs = {
```
How was this set compiled? Can we have this set shared somehow with the one in `load_sql_based_model`? I don't want to update one place and then forget about the other one

themisvaltinos replied:
right I will see how to either share or refactor to have a function with the same logic in both `load_sql_based_model` and here

izeigerman commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
+            "dialect": dialect,
+            "default_catalog": default_catalog,
+        }
+        meta_renderer = _meta_renderer(
```
Perhaps, instead of kwargs, the `_meta_renderer` function should take all the arguments above by name explicitly (no default values).

izeigerman commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
 
+        # Properties to be rendered at model creation time
+        expressions = [exp.Property(this="name", value=exp.to_table(self.name, dialect=dialect))]
+        for field_name in {"enabled", "start", "end"}:
```
Why only these properties? Shouldn't we instead take all fields from the `ModelMeta` schema with few exceptions like `signals`?

themisvaltinos replied:
ok I will add all fields of `ModelMeta` instead, aside from these that shouldn't be rendered at creation time. it will require specific handling for each property, but it can be done having something similar to `META_FIELD_CONVERTER` like in `render_definition`

georgesittas commented on sqlmesh/core/model/decorator.py:
```diff
@@ -12,9 +12,11 @@
+from sqlmesh.core import dialect as d
 from sqlmesh.core.model.definition import (
     Model,
+    _meta_renderer,
```
Should we rename this to get rid of the leading `_`? It's not a private helper anymore if we reference it here.

georgesittas commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
+        for field_name in {"enabled", "start", "end"}:
+            if (
+                (field_value := self.kwargs.get(field_name))
+                and isinstance(field_value, exp.Expression)
+                or (isinstance(field_value, str) and field_value.startswith(SQLMESH_MACRO_PREFIX))
```
Let's include parentheses so that operator precedence can be easily understood.

georgesittas commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
+                and isinstance(field_value, exp.Expression)
+                or (isinstance(field_value, str) and field_value.startswith(SQLMESH_MACRO_PREFIX))
+            ):
+                expressions.append(
+                    exp.Property(
+                        this=field_name,
+                        value=parse_one(field_value, dialect=dialect)
+                        if isinstance(field_value, str)
+                        else field_value,
+                    )
+                )
```
```suggestion
                expressions.append(
                    exp.Property(
                        this=field_name,
                        value=exp.maybe_parse(field_value, dialect=dialect),
                    )
                )
```

tobymao commented on sqlmesh/core/model/decorator.py:
```diff
@@ -119,21 +121,49 @@ def model(
         build_env(self.func, env=env, name=entrypoint, path=module_path)
 
+        # Properties to be rendered at model creation time
+        expressions = [exp.Property(this="name", value=exp.to_table(self.name, dialect=dialect))]
```
"name" should be a var, otherwise the ast is invalid

themisvaltinos commented:
Thanks for the review @izeigerman , @georgesittas , @tobymao  addressed comments and revised logic to handle all model properties, except for those that need to be rendered at evaluation time

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+    fields: t.Dict[str, t.Any],
+    module_path: Path,
+    path: Path,
+    jinja_macros: t.Optional[JinjaMacroRegistry] = None,
```
I suggest we don't have default values for any fields here, so that we know to explicitly provide them everywhere we call this

themisvaltinos replied:
The issue is they flow in as optional from model's arguments: 
https://github.com/TobikoData/sqlmesh/blob/dc613d05e9eb599fe564598143dd41429a8f72be/sqlmesh/core/model/decorator.py#L87-L90

izeigerman replied:
That's fine. I'm not suggesting making them non-optional. I suggest removing default arguments so that each caller consciously provide some value for every argument and doesn't miss them by accident.

themisvaltinos replied:
Oh I see I misunderstood revising this to remove the defaults

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+                    default_catalog=default_catalog,
+                )
+                assert rendered_expr and len(rendered_expr) == 1
+                return rendered_expr[0].sql(dialect=dialect)
```
This doesn't look right. If the input is expression, we return a string instead?

izeigerman replied:
Shouldn't we always return an expression?

themisvaltinos replied:
Yes revised it, this was needed for the model's name, but adjusted it for this field only

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+                assert rendered_expr and len(rendered_expr) == 1
+                return rendered_expr[0].sql(dialect=dialect)
+            except Exception:
+                pass
```
Why ignore?

themisvaltinos replied:
This was for cron that it was for example trying to render and failing to find `@daily` macro, but revised it and simply added cron in the  `RUNTIME_RENDERED_MODEL_FIELDS` as well.

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+        return value
+
+    for field_name, _ in ModelMeta.all_field_infos().items():
+        field_name = field_name[:-1] if field_name.endswith("_") else field_name
```
No need to do this. You can get the alias from the `FieldInfo` object in the value of the dict that you're iterating over.

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+    for field_name, _ in ModelMeta.all_field_infos().items():
+        field_name = field_name[:-1] if field_name.endswith("_") else field_name
+        if (field_value := fields.get(field_name)) and field_name not in {
+            "audits",
```
Should we set these in a named constant like `RUNTIME_RENDERED_MODEL_FIELDS`

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+            "virtual_properties",
+            "session_properties",
+        }:
+            if isinstance(field_value, t.Dict):
```
Hm, not sure whether it's a good practice to check for instance of `t.Dict`. I believe we just check for `dict` in other places.

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+        }:
+            if isinstance(field_value, t.Dict):
+                for key, value in field_value.items():
+                    if key != "merge_filter":
```
Can be a part of the same `RUNTIME_RENDERED_MODEL_FIELDS`

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+                for key, value in field_value.items():
+                    if key != "merge_filter":
+                        fields[field_name][key] = render_field_value(value)
+            elif isinstance(field_value, t.List):
```
Ditto: `isinstance(..., list)`

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,59 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+                    dialect=dialect,
+                    default_catalog=default_catalog,
+                )
+                assert rendered_expr and len(rendered_expr) == 1
```
Should we inform the user about the problem here?

themisvaltinos replied:
Right adapted it to follow instead the same pattern we have in the rest of the places

themisvaltinos commented on sqlmesh/core/constants.py:
```diff
@@ -90,3 +90,12 @@
+    "audits",
+    "signals",
+    "description",
+    "cron",
```
For description I was thinking we'd want to retain how it was unrendered, similarly to how we handle it here: https://github.com/TobikoData/sqlmesh/blob/c24ee34ff9cca73dcaba90b3e8780ac466c3c5d7/sqlmesh/core/model/definition.py#L201-L202
For cron since the user may pass a cron keyword like: `@hourly`, `@daily` trying to render it will fail since it will be considered a macro and unable to resolve it

themisvaltinos commented on sqlmesh/core/constants.py:
```diff
@@ -90,3 +90,12 @@
 
 DISABLE_SQLMESH_STATE_MIGRATION = "SQLMESH__AIRFLOW__DISABLE_STATE_MIGRATION"
+
+RUNTIME_RENDERED_MODEL_FIELDS = {
```
Right it's only referenced there for now so moved it

themisvaltinos commented on sqlmesh/core/model/definition.py:
```diff
@@ -2361,7 +2351,52 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+            if isinstance(field_value, dict):
+                for key, value in field_value.items():
+                    if key not in c.RUNTIME_RENDERED_MODEL_FIELDS:
+                        fields[field][key] = render_field_value(value)
```
Haven't considered this thanks revised it

themisvaltinos commented on sqlmesh/core/model/definition.py:
```diff
@@ -2382,7 +2381,62 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+            if rendered_expr is None:
+                raise SQLMeshError(
+                    f"Failed to render model: `{fields['name']}` at `{path}`\n"
+                    f"\tExpected rendering `{expression.sql(dialect=dialect)}` to return an expression"
```
Thanks for catching it my bad

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -2382,7 +2381,61 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
+            )
+            if rendered_expr is None:
+                raise SQLMeshError(
+                    f"Failed to render model: `{fields['name']}` at `{path}`\n"
```
```suggestion
                    f"Failed to render model attribute `{fields['name']}` at `{path}`\n"
```

izeigerman reviewed (approved):
A few smaller comments. Good to go otherwise!
